src/plugins/fs.c

```diff
@@ -817,3 +962,313 @@ gboolean bd_fs_xfs_resize (const gchar *mpoint, guint64 new_size, const BDExtraA
+            blkid_free_probe (probe);
+            synced_close (fd);
+            return FALSE;
+        }
+
+        ret->label = g_strdup (value);
+    }
+
+    status = blkid_probe_lookup_value (probe, "UUID", &value, NULL);
+    if (status != 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                     "Failed to get label for the device '%s'", device);
+        blkid_free_probe (probe);
+        synced_close (fd);
+        return FALSE;
+    }
+
+    ret->uuid = g_strdup (value);
+
+    blkid_free_probe (probe);
+    synced_close (fd);
+
+    success = bd_utils_exec_and_capture_output (args, NULL, &output, error);
+    if (!success)
+        /* error is already populated */
+        return FALSE;
+
+    lines = g_strsplit (output, "\n", 0);
+    g_free (output);
+    for (line_p=lines; *line_p && (!have_cluster_size || !have_cluster_count); line_p++) {
+        if (!have_cluster_size && g_str_has_suffix (*line_p, "bytes per cluster")) {
+            ret->cluster_size = g_ascii_strtoull (*line_p, NULL, 0);
+            have_cluster_size = TRUE;
+        } else if (!have_cluster_count && g_str_has_prefix (*line_p, device)) {
+            key_val = g_strsplit (*line_p, ",", 2);
+            sscanf (key_val[1], " %lu/%lu clusters", &full_cluster_count, &cluster_count);
+            ret->cluster_count = cluster_count;
+            ret->free_cluster_count = cluster_count - full_cluster_count;
+            have_cluster_count = TRUE;
+            g_strfreev (key_val);
+        }
+    }
+    g_strfreev (lines);
+
+    return ret;
+}
+
+/**
+ * init: (skip)
+ */
+gboolean init() {
+    ped_exception_set_handler ((PedExceptionHandler*) exc_handler);
+    return TRUE;
+}
+
+/**
+ * set_parted_error: (skip)
+ *
+ * Set error from the parted error stored in 'error_msg'. In case there is none,
+ * the error is set up with an empty string. Otherwise it is set up with the
+ * parted's error message and is a subject to later g_prefix_error() call.
+ *
+ * Returns: whether there was some message from parted or not
+ */
+static gboolean set_parted_error (GError **error, BDFsError type) {
+    gchar *error_msg = NULL;
+    error_msg = get_error_msg ();
+    if (error_msg) {
+        g_set_error (error, BD_FS_ERROR, type,
+                     " (%s)", error_msg);
+        g_free (error_msg);
+        error_msg = NULL;
+        return TRUE;
+    } else {
+        g_set_error_literal (error, BD_FS_ERROR, type, "");
+        return FALSE;
+    }
+}
+
+/**
+ * bd_fs_vfat_resize:
+ * @device: the device the file system of which to resize
+ * @new_size: new requested size for the file system (if 0, the file system is
+ *            adapted to the underlying block device)
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the file system on @device was successfully resized or not
+ */
+gboolean bd_fs_vfat_resize (const gchar *device, guint64 new_size, GError **error) {
```

Comment:
This function needs to take care about progress reporting on it's own as it doesn't run any utility.